cliquet/utils.py

```diff
@@ -116,3 +144,46 @@ def current_service(request):
         pattern = request.matched_route.pattern
         service = services[pattern]
         return service
+
+
+def cached(ttl=None, prefix='cached_'):
+    """Decorator to cache the result of a function or method into the
+    configured cache backend.
+
+    :param float ttl: expire after number of seconds.
+    :param str prefix: prefix cache keys with specific string.
+
+    .. code-block:: python
+
+        EXPIRATION = 3600  # seconds.
+
+        class Permissions(object):
+            @cached(ttl=EXPIRATION)
+            def has_perms(self, user_id):
+                resp = requests.get('http://server/?allowed=' % user_id)
+                return resp.status_code == 200
+
+    """
+    from pyramid.threadlocal import get_current_registry
+
+    # On class method, skip ``self`` to compute cache key.
+    frames = inspect.stack()
+    defined_in_class = (len(frames) > 2 and
+                        frames[2][4][0].strip().startswith('class '))
+    skip_args = 1 if defined_in_class else 0
+
+    # Get cache from current thread.
+    cache = get_current_registry().cache
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapped(*args, **kwargs):
+            key = prefix + func.__name__ + str(args[skip_args:]) + str(kwargs)
```

Comment:
I remember that @Natim did a great hashing function for the cache keys in some project but can't remember where...